pkg/core/inject/context.go

```diff
@@ -0,0 +1,72 @@
+package inject
+
+import (
+	"context"
+)
+
+type contextKeyType struct{}
+
+var contextKey = contextKeyType{}
+
+type injectContext struct {
+	Root         *injectContext
+	Parent       *injectContext
+	Dependencies map[string]interface{}
+}
+
+func WithInject(ctx context.Context) context.Context {
+	injectCtx := &injectContext{Dependencies: map[string]interface{}{}}
+
+	if baseCtx, ok := ctx.Value(contextKey).(*injectContext); ok {
+		injectCtx.Parent = baseCtx
+		injectCtx.Root = baseCtx.Root
+	} else {
+		injectCtx.Parent = nil
+		injectCtx.Root = injectCtx
```

Comment:
We design the inject context to support getting the same dependency from Root by using `Singleton` function, but we haven't used it, right? Since we called `WithInject ` in `DefaultRequestInject`, but the context is used within the `DefaultRequestInject` only. If this is the case, maybe leave some comment in `DefaultRequestInject` function?

Reply:
In our current usage:
- `Transient` and `Scoped` would work correctly.
- `Singleton` would act same as `Scoped`, since the request does not have a parent inject context. We shouldn't use it for now.
- Context of request is not updated with the inject context. Handlers should not use these function with request context.

I think I'll add comments to the functions to indicate the current limitations.

Reply:
I guess parent is not being used now?

Reply:
Yes, I was thinking whether scoped dependencies should be resolved from all contexts in the chain, or just the current context. For simplicity, it is resolved in current context at the moment.